### DISCUSSION TITLE
Formatting fields into correct form

### DISCUSSION THREAD

**Discussion by DavidKmn** [pointfreeco/swift-url-routing#36] [Q&A] [answered]
So I am in the process of getting started with the library and am a bit confused about how we could map on the fields to convert them to a usable format. So here is a situation where I want to uppercase the from wallet value before initialising my enum as the enum has an uppercase raw value requirement. Is this possible? Tried to look at the docs but with no luck.

![Screenshot 2022-06-29 at 06 27 38](https://user-images.githubusercontent.com/35953014/176358749-d31f04b3-7562-474a-baae-a1d02c2e43fc.png)

Full Error: `Initializer 'init(_:with:)' requires that 'Parsers.Map<Field<Parsers.MapConversion<Parsers.ReplaceError<Rest<Substring>>, Conversions.SubstringToString>>, String>' conform to 'ParserPrinter'`

**Comment by robfeldmann** [comment, marked as answer]
I think this is because `map` by itself isn't enough for parsing **and** printing. You will have to define how to go both ways. 

There's a special conversation you can use. Something like:

```
.map(.convert(
  apply: {
    // for parsing
    $0.uppercased()
  },
  unapply: {
    // for printing
    $0.lowercased()
}
))
```

**Comment by mbrandonw** [comment]
Yep, @robfeldmann is right!

You can also extract it out to an extension if doing all of that inline is too intense:

```swift
extension Conversion where Self == AnyConversion<Substring, Substring>  {
  static var uppercase: AnyConversion<Input, Output> {
    AnyConversion(apply: { $0.uppercased()[...] }, unapply: { $0.uppercased()[...] })
  }
}
```

Then you can use it like so: 

```swift
Field("fromWallet").map(.uppercase)
```

Also note that what you put in the `unapply` closure is what is used for printing back to a URL. So you could choose to uppercase, lowercase, or leave it fixed with `{ $0 }`.

**Reply by DavidKmn** [reply]
Awesome, thanks for the help, this is exactly what I needed!
